src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -0,0 +1,291 @@
+  actor_table_data->CopyFrom(GetActorTableData());
+
+  auto actor = shared_from_this();
+  auto on_done = [actor, callback](Status status) {
+    if (!status.ok()) {
+      // TODO(zsl): flush with retry?
+      RAY_LOG(FATAL) << "Failed to update the state of actor " << actor->GetActorID()
+                     << " to " << ActorTableData::ActorState_Name(actor->GetState())
+                     << " with node id " << actor->GetNodeID();
+    }
+    RAY_LOG(INFO) << "Succeeded in updating the state of actor " << actor->GetActorID()
+                  << " to " << ActorTableData::ActorState_Name(actor->GetState())
+                  << " with node id " << actor->GetNodeID();
+    if (callback) {
+      callback();
+    }
+  };
+  auto status =
+      redis_gcs_client_->Actors().AsyncUpdate(GetActorID(), actor_table_data, on_done);
+  if (!status.ok()) {
+    on_done(status);
+  }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+GcsActorManager::GcsActorManager(boost::asio::io_context &io_context,
+                                 std::shared_ptr<gcs::RedisGcsClient> gcs_client,
+                                 gcs::GcsNodeManager &gcs_node_manager)
+    : redis_gcs_client_(std::move(gcs_client)),
+      gcs_actor_scheduler_(io_context, gcs_node_manager) {
+  RAY_LOG(INFO)
+      << "Initializing GcsActorManager and load actor information from storage.";
+  auto actor_table_data_list = LoadAllActorTableData();
+  for (auto &actor_table_data : actor_table_data_list) {
+    auto actor =
+        std::make_shared<GcsActor>(std::move(actor_table_data), redis_gcs_client_);
+    // Ignore dead actor.
+    if (actor->GetState() != rpc::ActorTableData::DEAD) {
+      actor->UpdateState(rpc::ActorTableData::PENDING);
+      registered_actors_.emplace(actor->GetActorID(), actor);
+      pending_actors_.emplace(actor);
+    }
+  }
+
+  gcs_node_manager.AddNodeAddedListener(
+      [this](const std::shared_ptr<rpc::GcsNodeInfo> &) {
+        // Because a new node has been added, we need to try to schedule the pending
+        // actors.
+        SchedulePendingActors();
+      });
+
+  gcs_node_manager.AddNodeRemovedListener([this](std::shared_ptr<rpc::GcsNodeInfo> node) {
+    // All of the related actors should be reconstructed when a node is remove from gcs
+    // node manager.
+    ReconstructActorsOnNode(ClientID::FromBinary(node->node_id()));
+  });
+
+  gcs_actor_scheduler_.SetScheduleFailedHandler([this](std::shared_ptr<GcsActor> actor) {
+    // When there are no available nodes to schedule the actor the gcs_actor_scheduler
+    // will treat it as failed and invoke this handler. In this case, the actor should be
+    // pended to the `pending_actors_` and wait for the registion of new node.
+    pending_actors_.emplace(std::move(actor));
+  });
+
+  gcs_actor_scheduler_.SetScheduleSuccessHandler([this](std::shared_ptr<GcsActor> actor) {
```

Comment:
maybe make the names in pair? e.g.  `Success` and `Failure`, or `Succeeded` and `Failed`